Affiliate server

onvej-sl opened the pull request:
This pull request implements communication with affiliate servers. @lontivero, please take a look at it.

### Brief overview of architecture

`AffiliateInformation` is published in the status of the coordinator (https://github.com/zkSNACKs/WalletWasabi/commit/0cc9f6ad7b0d63735e11ade4e4de0bc5d86f2068). The information is supplied by `AffiliationManager` (https://github.com/zkSNACKs/WalletWasabi/commit/b70b28f47e6d4416501593cd7ea05a9290add2a7) that is running as a background service (https://github.com/zkSNACKs/WalletWasabi/commit/e9332a7b82e59f944de28fd6d41068f16828ad4f).

`AffiliateInformation` (https://github.com/zkSNACKs/WalletWasabi/commit/65af5c28d334681365efed685cbabc3a116dc271) consists of the following two fields:
  * A list of running affiliate servers. 
    * A client that affiliates a server that is not running won't join a coinjoin round. If they did, they would ruin the round for all the participants, because they wouldn't be able to sign the final coinjoin transaction.
    * The list of running affiliate servers is supplied by `AffiliateServerStatusUpdater`.
      * The `AffiliateServerStatusUpdater` (https://github.com/zkSNACKs/WalletWasabi/commit/739d3732af552277951a4327a0d7d28c0ea34435) is a service that is periodically trying to connect to the affiliate servers mentioned in the configuration (https://github.com/zkSNACKs/WalletWasabi/commit/7c13c295ff8657fb154554dff63f3492b0d59032) to determine which affiliate servers are running.
  * Coinjoin requests for each round that is in the signing phase and for each affiliate server.
    * A coinjoin request is a piece of data used by hardware wallets to validate and sign the final coinjoin transaction.
    * The coinjoin requests are supplied by `CoinjoinRequestsUpdater`.
      * `CoinjoinRequestsUpdater` (https://github.com/zkSNACKs/WalletWasabi/commit/37388ef0dab047e8cf1d8415a3b165e5e907fde8) collects information about ongoing rounds. `Arena` emits an event if a round is created, a round changes their phase or a final coinjoin transaction is ready to be signed (https://github.com/zkSNACKs/WalletWasabi/commit/12679536f03f77ba0cde5583bf5b50c9c7d051af and https://github.com/zkSNACKs/WalletWasabi/commit/293f47fbf1e4877b601e909c938684acec080721). Clients send their affiliation in the `ready-to-sign` request (https://github.com/zkSNACKs/WalletWasabi/commit/66c47e1f330517906bb0d6b00ba7c185bc43b50b). All this information is stored in `RoundData` (https://github.com/zkSNACKs/WalletWasabi/commit/7f063f83eaca7c88c30203b10326ef380f9e9f2e).

The coordinator communicate with an affiliate server using json HTTP POST requests (https://github.com/zkSNACKs/WalletWasabi/commit/08d5c17455e1003fb18c5dbeb6ce28f81be35886). An affiliate server implements two methods:
 * `get_status` is used to determine whether an affiliate server is running.
 * `get_coinjoin_request` is used to acquire a coinjoin request. The format of the `get_coinjoin_request` request is implemented in https://github.com/zkSNACKs/WalletWasabi/commit/e112eb36208f59915e0782bb01581ac9f35a9241. To ensure authenticity, the request is first serialized to a canonical json format (https://github.com/zkSNACKs/WalletWasabi/commit/f7a5ff4cf29419ecb8a982a1db0757d844d7f2f2) and then signed (https://github.com/zkSNACKs/WalletWasabi/commit/00f3c521c03d914a8920d868a0376479ad617ff0).

lontivero commented on WalletWasabi/Affiliation/AffiliateInformationUpdater.cs:
```diff
@@ -0,0 +1,297 @@
+		Body body = roundData.GetAffiliationData(affiliationFlag);
+		byte[]? result = await GetCoinjoinRequestAsync(affiliateServerHttpApiClient, body);
+
+		if (result is not null)
```
Do not use null as a way to indicate that something was wrong because we lose the information about what was wrong. The the exception to bubble and catch it here were we can log the problem.

onvej-sl replied:
Done in https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/0b6565315458c4fc1e1b4335ac2ef357cddb7b38.

lontivero commented on WalletWasabi/Affiliation/AffiliateInformationUpdater.cs:
```diff
@@ -0,0 +1,297 @@
+
+			Dictionary<AffiliationFlag, byte[]> coinjoinRequests = CoinjoinRequests[roundId];
+
+			if (coinjoinRequests.ContainsKey(affiliationFlag))
```
Avoid tester-doer pattern and use TryAdd instead

onvej-sl replied:
Done in https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/0b6565315458c4fc1e1b4335ac2ef357cddb7b38.

lontivero commented on WalletWasabi/Affiliation/AffiliateInformationUpdater.cs:
```diff
@@ -0,0 +1,297 @@
+			throw new InvalidOperationException($"The round ({roundId}) does not exist.");
+		}
+
+		if (CoinjoinRequests.ContainsKey(roundId))
```
Avoid tester-doer pattern and use TryRemove

onvej-sl replied:
As far as I know, `Dictionary` doesn't implement `TryRemove`, see [this](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.dictionary-2?view=net-7.0).

onvej-sl replied:
Done in https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/76e434c939215dbff81a865ba3091f1afafa2ec0.

lontivero reviewed (changes requested):
I don't want to overwhelm with suggestions/comments/change requests so, this is a very partial review. I will continue batching suggestions in small chunks.

lontivero commented on WalletWasabi/Affiliation/AffiliateInformationUpdater.cs:
```diff
@@ -0,0 +1,297 @@
+	{
+		if (phase == Phase.Ended)
+		{
+			RoundsToRemove.Enqueue(roundId);
```
Why to enque and not simply remove it?

onvej-sl replied:
If I remember it correctly, the only reason was that I wanted to have the event handler as simple as possible because it is called synchronously. Maybe it is an overkill at this place.

lontivero replied:
Yes, imo just removing the round would simplify the code.

onvej-sl replied:
Done in https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/cb5525e98eabc24ed5655b2520d1d17642312281.

lontivero commented on WalletWasabi/Affiliation/IAffiliationManager.cs:
```diff
@@ -0,0 +1,8 @@
+
+namespace WalletWasabi.Affiliation;
+
+public interface IAffiliationManager
```
It seems this interface is not necessary. Please remove it.

onvej-sl replied:
The interface is used in tests. See [this](https://github.com/trezor/WalletWasabi/blob/ddba48534fbc090f20e577ac82e42abc0ab7cfdc/WalletWasabi.Tests/UnitTests/WabiSabi/Integration/WabiSabiApiApplicationFactory.cs#L68).

lontivero replied:
That's because the AffiliationManager doesn't receive its dependencies and then you have to replace it completely with a fake/mock instance. 

In UTs we have to break dependencies with the external world, in this case we have to break dependency with the network so, you could receive a factory method that creates IHttpClient.

onvej-sl replied:
> That's because the AffiliationManager doesn't receive its dependencies and then you have to replace it completely with a fake/mock instance.

The question is whether you want to use the real `AffiliationManager` in tests if it is not used by anyone.

> In UTs we have to break dependencies with the external world, in this case we have to break dependency with the network so, you could receive a factory method that creates IHttpClient.

Sorry, I don't understand what you mean. I feel this is closely related to https://github.com/zkSNACKs/WalletWasabi/pull/9769#discussion_r1048694574. So I suggest to solve that issue first.

lontivero replied:
> The question is whether you want to use the real AffiliationManager in tests 

Yes. That's the idea.

onvej-sl replied:
Is this what you expected? https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/719c21d2c87c6eafe7f0096245b221c85b8d66f6

lontivero replied:
yes.

lontivero reviewed (changes requested):
Second shallow review.

nopara73 reviewed (changes requested):
I am reiterating what I've said many times before. An affiliate server cannot be allowed to disrupt coinjoin rounds. You may prevent inputs from being accepted for registration into coinjoins, but it is not acceptable to have the ability to disrupt rounds. In other words affiliate servers cannot play after the input registration has concluded.

nopara73 commented:
Here's an alternative solution:

During fresh bitcoin input registration the WW coord sends a request to the TR aff with the input (and whatever else you want to verify.) TR aff replies with a signature that is forwarded to the TR client. TR client verifies and decides to go forward or not.

onvej-sl commented on WalletWasabi/Affiliation/AffiliationFlag.cs:
```diff
@@ -0,0 +1,64 @@
+namespace WalletWasabi.Affiliation;
+
+[TypeConverter(typeof(AffiliationFlagConverter))]
+public class AffiliationFlag : IEquatable<AffiliationFlag>
```
Done in https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/fdd598f57098815ada1fd1e51126a2e8d3c4a9b3.

onvej-sl commented on WalletWasabi/Affiliation/AffiliationFlag.cs:
```diff
@@ -0,0 +1,64 @@
+public class AffiliationFlag : IEquatable<AffiliationFlag>
+{
+	public static AffiliationFlag Default = new AffiliationFlag("WalletWasabi");
+	public static AffiliationFlag Trezor = new AffiliationFlag("trezor");
```
I don't think they should be:
  *  `Trezor` is used only in tests, I can inline it if you wish.
  * `WalletWasabi` is the affiliation flag of your users. There is no reason to have it configurable. Actually, the value is not send anywhere. The only thing that matters is that this value is different from all other flags.
  * There is no need to add a new affiliation flag here if a new affiliate server is created.

lontivero replied:
ok

onvej-sl commented on WalletWasabi/Affiliation/Serialization/AffiliationFlagJsonConverter.cs:
```diff
@@ -0,0 +1,30 @@
+		}
+		else
+		{
+			throw new Exception();
```
> Btw, in most of the cases we return null here because that's the expected behavior according to the docs and the method signature (AffiliationFlag?)

I want the serializer to be strict and don't allow it to return `null`. I concluded `JsonSerializationException` is the most appropriate.  See https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/4408ad25fb9579305f1c8a27b61f2be6e22b0b24.

onvej-sl commented on WalletWasabi/Affiliation/Serialization/AffiliationFlagJsonConverter.cs:
```diff
@@ -0,0 +1,30 @@
+		{
+			throw new ArgumentNullException(nameof(value));
+		}
+		else
```
Done in https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/4408ad25fb9579305f1c8a27b61f2be6e22b0b24.

onvej-sl commented on WalletWasabi/Affiliation/Models/AffiliateInformation.cs:
```diff
@@ -0,0 +1,15 @@
+	IReadOnlyDictionary<uint256, IReadOnlyDictionary<AffiliationFlag, byte[]>> CoinjoinRequests
+)
+{
+	public static AffiliateInformation Empty = new AffiliateInformation(Enumerable.Empty<AffiliationFlag>(), ImmutableDictionary<uint256, IReadOnlyDictionary<AffiliationFlag, byte[]>>.Empty);
```
Done in https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/bbd13ba858a8c1e0013e2fae9424bd7e59a1726a.

onvej-sl commented on WalletWasabi/Affiliation/AffiliateInformationUpdater.cs:
```diff
@@ -0,0 +1,297 @@
+
+	protected override async Task ActionAsync(CancellationToken cancel)
+	{
+		await UpdateCoinjoinRequestsAsync();
```
Done in https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/0b6565315458c4fc1e1b4335ac2ef357cddb7b38.

onvej-sl commented on WalletWasabi/Affiliation/AffiliateInformationUpdater.cs:
```diff
@@ -0,0 +1,297 @@
+
+	private async Task UpdateCoinjoinRequestsAsync()
+	{
+		while (RoundsToUpdate.Any())
```
Done in https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/0b6565315458c4fc1e1b4335ac2ef357cddb7b38.

onvej-sl commented on WalletWasabi/Affiliation/AffiliateInformationUpdater.cs:
```diff
@@ -0,0 +1,297 @@
+
+	private async Task UpdateCoinjoinRequestsAsync(uint256 roundId)
+	{
+		if (!RoundData.TryGetValue(roundId, out RoundData? roundData))
```
Done in https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/0b6565315458c4fc1e1b4335ac2ef357cddb7b38.

onvej-sl commented on WalletWasabi/Affiliation/AffiliateInformationUpdater.cs:
```diff
@@ -0,0 +1,297 @@
+		}
+	}
+
+	private async Task UpdateCoinjoinRequestsAsync(uint256 roundId)
```
Done in https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/0b6565315458c4fc1e1b4335ac2ef357cddb7b38.

onvej-sl commented on WalletWasabi/Affiliation/AffiliateInformationUpdater.cs:
```diff
@@ -0,0 +1,297 @@
+	private async Task UpdateCoinjoinRequestsAsync(uint256 roundId, RoundData roundData, AffiliationFlag affiliationFlag, AffiliateServerHttpApiClient affiliateServerHttpApiClient)
+	{
+		Body body = roundData.GetAffiliationData(affiliationFlag);
+		byte[]? result = await GetCoinjoinRequestAsync(affiliateServerHttpApiClient, body);
```
Done in https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/0b6565315458c4fc1e1b4335ac2ef357cddb7b38.

onvej-sl commented on WalletWasabi/Affiliation/RoundData.cs:
```diff
@@ -0,0 +1,113 @@
+		Transaction = null;
+	}
+
+	private RoundParameters? RoundParameters { get; set; }
```
Done in https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/cdb3a2fd79e6b8b966054110c03c91d2a470201d.

onvej-sl commented on WalletWasabi/Affiliation/AffiliateServerStatusUpdater.cs:
```diff
@@ -0,0 +1,96 @@
+	private static TimeSpan AffiliateServerTimeout = TimeSpan.FromSeconds(20);
+	private static TimeSpan Interval = TimeSpan.FromMinutes(1);
+
+	public AffiliateServerStatusUpdater(IDictionary<AffiliationFlag, AffiliateServerHttpApiClient> clients)
```
Good to know! Can you suggest better name that `Interval`?

lontivero replied:
No no, interval is okay i think. This is just to share some knowledge that could be important in some scenarios (not in this case it seems)

onvej-sl commented on WalletWasabi/Affiliation/AffiliateServerStatusUpdater.cs:
```diff
@@ -0,0 +1,96 @@
+
+	protected override async Task ActionAsync(CancellationToken cancel)
+	{
+		await UpdateRunningAffiliateServersAsync();
```
Done in https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/1a8b193a177855f25bcc4578df5a4af2adec018f.

onvej-sl commented on WalletWasabi/Affiliation/AffiliateServerStatusUpdater.cs:
```diff
@@ -0,0 +1,96 @@
+				}
+			}
+		}
+		catch (Exception exception)
```
Done in https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/1a8b193a177855f25bcc4578df5a4af2adec018f.

onvej-sl commented on WalletWasabi/Affiliation/Serialization/FeeRateJsonConverter.cs:
```diff
@@ -0,0 +1,44 @@
+		}
+		else
+		{
+			throw new Exception();
```
Done in https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/4acfcb585bfce8028511e1c58f86f8b77f2e38dd but see https://github.com/zkSNACKs/WalletWasabi/pull/9769#discussion_r1047577920.

onvej-sl commented on WalletWasabi/WabiSabi/Backend/Rounds/Arena.cs:
```diff
@@ -632,4 +637,22 @@ private Script GetCoordinatorScriptPreventReuse(Round round)
+	private void AddRound(Round round)
+	{
+		Rounds.Add(round);
+		RoundCreated?.Invoke(this, new RoundCreatedEventArgs(round.Id, round.Parameters));
```
Done in https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/3ea9339b9796b5a1735e024ce0083d8d272e1b16.

onvej-sl commented on WalletWasabi/WabiSabi/Backend/Rounds/Arena.Partial.cs:
```diff
@@ -12,11 +12,14 @@
 
 public partial class Arena : IWabiSabiApiRequestHandler
 {
+	public event EventHandler<AffiliationAddedEventArgs>? AffiliationAdded;
```
Done in https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/3ea9339b9796b5a1735e024ce0083d8d272e1b16.

onvej-sl commented on WalletWasabi/WabiSabi/Backend/Rounds/Arena.Partial.cs:
```diff
@@ -129,6 +132,7 @@ public async Task ReadyToSignAsync(ReadyToSignRequestRequest request, Cancellati
 			var round = GetRound(request.RoundId, Phase.OutputRegistration);
 			var alice = GetAlice(request.AliceId, round);
 			alice.ReadyToSign = true;
+			AffiliationAdded?.Invoke(this, new AffiliationAddedEventArgs(round.Id, alice.Coin, request.AffiliationFlag, alice.IsPayingZeroCoordinationFee));
```
Is this what you mean?

```
private void NotifyAffiliation(uint256 roundId, Coin coin, AffiliationFlag affiliationFlag, bool isPayingZeroCoordinationFee)
	{
		AffiliationAdded?.SafeInvoke(this, roundId, coin, affiliationFlag, isPayingZeroCoordinationFee);
	}
```

I have no problem doing it but I don't see the benefit of doing it this way.

lontivero replied:
Yes, something like that. 
```c#
private void NotifyAffiliation(uint256 roundId, Coin coin, AffiliationFlag affiliationFlag, bool isPayingZeroCoordinationFee)
{
	AffiliationAdded.SafeInvoke(new AffiliationAddedEventArgs(round.Id, alice.Coin, request.AffiliationFlag, alice.IsPayingZeroCoordinationFee));
}
```

------

`Arena.cs` and `Arena.partial.cs` are the same logical unit, the same class, and that means that it really doesn't matter which file you decide to touch the result is the same. However, there is a clear separation in terms of what the code in each of those two files is for:

* Arena.cs contains the "engine", the worker moves the rounds from phase to phase and does almost everything while,
* Arena.cs is simply the entry point to Arena, the one that communicates with the external world and it is almost a perfect 1-to-1 with the HTTP API. 

So, it is okay to raise the event from here but it is unexpected. I would keep Arena.Partial.cs clean from this concepts of events and would put all the event definitions and raising in Arena.cs.

onvej-sl replied:
Ok, I think understand. Done in https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/013eb199e694996329c167f5afdb3fe0b750a091.

onvej-sl commented on WalletWasabi/WabiSabi/Backend/Rounds/Arena.cs:
```diff
@@ -274,6 +278,7 @@ private async Task StepTransactionSigningPhaseAsync(CancellationToken cancellati
 				if (state.IsFullySigned)
 				{
 					Transaction coinjoin = state.CreateTransaction();
+					CoinjoinTransactionCreated?.Invoke(this, new CoinjoinTransactionCreatedEventArgs(round.Id, coinjoin));
```
I'm not sure If I understand. Is this what you mean?
```
	private Transaction CreateTransaction(Round round)
	{
		SigningState state = round.Assert<SigningState>();
		Transaction coinjoin = state.CreateTransaction();
		CoinjoinTransactionCreated?.SafeInvoke(this, new CoinjoinTransactionCreatedEventArgs(round.Id, coinjoin));
		return coinjoin;
	}
```

> Also, I would use SafeInvoke everywhere.

Done in https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/3ea9339b9796b5a1735e024ce0083d8d272e1b16.

lontivero replied:
Yes, something like that. Perhaps I would pass the SigningState instead of the Round but you got the idea.

onvej-sl replied:
> Perhaps I would pass the SigningState instead of the Round but you got the idea.

That's not possible because the method has to know round id.

Which one do you prefer?
```c#
private Transaction CreateTransaction(Round round)
```
```c#
private Transaction CreateTransaction(uint256 roundId, SigningState signingState)
```

lontivero replied:
You're right. Both are okay to me but the second one is more correct to me.

onvej-sl replied:
Done in https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/e31340c4250f3a129cb0c46102a155550d493ba7.

onvej-sl commented on WalletWasabi/Affiliation/Signer.cs:
```diff
@@ -0,0 +1,32 @@
+
+public class Signer
+{
+	private ECDsa ECDsa;
```
Done in https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/f0210fa282b6240c3a570e628156722e501cd020.

lontivero replied:
Sadly the problem with Disposables is the same that happens with async, lazy, purity, etc. Once you have a reference to a disposable instance, everything is disposable all the way up. In this case the AffiliationServer must also be IDisposable and you have to dispose it explicitly by calling its Dispose method.

onvej-sl replied:
I though I did it but apparently I didn't finish it.

 Done in https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/6a346adfd880120999a8a8d0461e99d92babd74d and  https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/e26a3eead0fae584b2a5ad76220b3aa0daa7557c.

onvej-sl commented on WalletWasabi/Affiliation/IAffiliateServerApi.cs:
```diff
@@ -0,0 +1,12 @@
+
+namespace WalletWasabi.Affiliation;
+
+public interface IAffiliateServerApi
```
I have no problem removing it. I agree it is not necessary but it clearly documents the API of the affiliate server. What do you think?

lontivero replied:
We make an effort to avoid interfaces as much as possible. Many times we were tempted to use them for unit testing or breaking some dependencies but we realize it is possible (and good) to don't have them. We have never used interfaces for documenting, i know that's a common and desirable thing in many languages (i am thinking in ocaml) but we don't do it and i think we don't need it neither.

onvej-sl replied:
Ok. Done in https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/6cbfa6ec17b955f718aa2ba8efe18b51bd16c4f3.

onvej-sl commented:
@lontivero Thank you for your super fast review, we appriciate it! I addressed most of your comments, I'm going to do the rest tomorrow. Please use "Resolve conversation" to make the list of unsolved issues more clearer.

onvej-sl commented:
> An affiliate server cannot be allowed to disrupt coinjoin rounds.

It's not an affiliate server that is disrupting rounds. A round can be only disrupted by a user regardless whether they use an affiliate server or not. Trezor users won't be participating in rounds if the Trezor affiliate server is down.

> In other words affiliate servers cannot play after the input registration has concluded.

The Trezor affiliate server cannot give its users consent to sign the coinjoin transaction until it sees it.

> During fresh bitcoin input registration the WW coord sends a request to the TR aff with the input (and whatever else you want to verify.) TR aff replies with a signature that is forwarded to the TR client. TR client verifies and decides to go forward or not.

I don't understand what problem it solves.

@andrewkozlik @MaxHillebrand

nopara73 commented:
> The Trezor affiliate server cannot give its users consent to sign the coinjoin transaction until it sees it.

It can actually. You don't need to have the exact transaction, you just need to know the *kind of transaction* you allow your users to participate in. And that permission can be given out at input registration, no need to introduce a denial of service attack vector for that.

andrewkozlik commented:
> During fresh bitcoin input registration the WW coord sends a request to the TR aff with the input (and whatever else you want to verify.) TR aff replies with a signature that is forwarded to the TR client. TR client verifies and decides to go forward or not.

> You don't need to have the exact transaction, you just need to know the kind of transaction you allow your users to participate in.

I don't agree. In our security model the desktop software is untrusted, i.e. the system must be secure even if there is malware on the user's computer interacting with Trezor. It's important for the affiliate server to confirm the transaction as a whole. If the affiliate's signature in the CoinJoin request would only be bound to one input UTXO, then it can be used to make Trezor sign a transaction which was not created by the coordinator and which transfers the coordination fee for that input UTXO to an attacker's address. So the user is not getting to participate in a CoinJoin, they lose out on paying the mining fee and the coordination fee goes to the attacker. This process can be repeated with the user's output from the malicious transaction. The malware will again register that output for CoinJoin and receive a signed CoinJoin request from the affiliate server. Since the coordinator/affiliate does not recognize the output as coming from a CoinJoin, the CoinJoin request will be instructing Trezor to pay the coordination fee again. The malware will again create a malicious transaction transferring the fee to the attacker.

Since the CoinJoin process on Trezor is designed to be unattended, by the time the user comes back, their entire account balance may be gone on mining fees and coordination fees.

onvej-sl commented on WalletWasabi/Affiliation/Signer.cs:
```diff
@@ -0,0 +1,32 @@
+
+namespace WalletWasabi.Affiliation;
+
+public class Signer
```
It's an affiliate server who signs coinjoin requests. `Signer` signs `CoinJoinRequestRequest` so the most concise name would be `CoinJoinRequestRequestsSigner` which is very cumbersome. Maybe we should rename coinjoin requests to something else.

onvej-sl commented on WalletWasabi/WabiSabi/Backend/WabiSabiConfig.cs:
```diff
@@ -164,6 +166,14 @@ public WabiSabiConfig(string filePath) : base(filePath)
 
+	[DefaultValue(Constants.FallbackCoordinatorSignerKeyHex)]
+	[JsonProperty(PropertyName = "CoordinatorSignerKeyHex", DefaultValueHandling = DefaultValueHandling.Populate)]
+	public string CoordinatorSignerKeyHex { get; set; } = Constants.FallbackCoordinatorSignerKeyHex;
```
What do you suggest?

lontivero replied:
The problem with Signer is that it is too generic and even with some context it is not immediately clear what it is for. 

In this case, while the Affiliation Server is authenticated with a certificate, the coordinator uses this as a poor's man authentication mechanism. I say poor's man because the most common way to do this kind of things is by pushing the authentication to the infrastructure level and use a certificate on the coordinator's side too.

I think it worth to come back to this later. Imo we should have to have strong reasons to implement things like this at application level.

In the meanwhile a better name could be something like AffiliationMessageSigner or similar.

onvej-sl replied:
> In this case, while the Affiliation Server is authenticated with a certificate, the coordinator uses this as a poor's man authentication mechanism. I say poor's man because the most common way to do this kind of things is by pushing the authentication to the infrastructure level and use a certificate on the coordinator's side too.

> I think it worth to come back to this later. Imo we should have to have strong reasons to implement things like this at application level.

I don't agree with you:
  * This has nothing to do with quality of the code. It is a matter of the protocol between the coordinator and the affiliation server. Moving the authentication to the transport layer changes the protocol.
  * The reason why the protocol is designed in this way is we want the affiliate server to have a short proof of what the coordinator requested without having to remember the whole communication in transport layer.
  * This approach is not definitely not unusual. [Ownership proofs](https://github.com/satoshilabs/slips/blob/master/slip-0019.md), for example, use a very similar concept.

> In the meanwhile a better name could be something like AffiliationMessageSigner or similar.

I don't think so. To be precise, `CoordinatorSignerKeyHex` is a private (signing) key of a coordinator that is first DER encoded (to bytes) and then hex encoded (to string).

lontivero replied:
> The reason why the protocol is designed in this way is we want the affiliate server to have a short proof of what the coordinator requested without having to remember the whole communication in transport layer.

If you move authentication to the transport layer you know that whatever reaches your server is authentic. You can then log the requests (not the traffic).

Otherwise, why do you think it is important for you to have a proof of what the coordinator requested but it is not important for the coordinator to have a proof of what the affiliation server to responded? Should the affiliation client know the affiliation server public key to be able to verify the signature of the coinjoinRequest response?

> I don't think so. To be precise, CoordinatorSignerKeyHex is a private (signing) key of a coordinator that is first DER encoded (to bytes) and then hex encoded (to string).


Yes, sorry, i was talking about the Signer class and not the private key here. Once you find a better name for the Signer the better name for this will become easier to find.

lontivero replied:
Please change the name. I will not block this for a silly renaming but it would be good to have a name that help us to understand what we are reading. Make some reference to the purpose of the key at least, here my ugly try: AffiliationMessageSigner for the signer and AffiliationMessageSignerKey for the key.

onvej-sl replied:
Done in https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/80baeacf449ed871d3bdd8f89950a5805c3dc0b3, https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/4edcf80dedba16cfb9f3599f98ed7b5d4fa919f4 (and https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/9b512968e7632b5f6f2ef078049ad7e34b508269, https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/dc39f3debad8e95c6f40f30498fbb8235a59a4c8, https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/38ea11994292722d4fab2e38ec9ac2c450f35987, https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/c3ba0079078564790dc0737e014b283395457a26, https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/ab9b5dc093b8fe9325ba0b58fcfb119cc4d62a80).

andrewkozlik commented:
> An affiliate server cannot be allowed to disrupt coinjoin rounds. 

We have been testing the solution for around two months now, and the affil server never caused any problems. The main reasons for CoinJoins failing is due to Tor timeouts and Trezor accidentally being disconnected just before signing.

In case the affiliate server happens to be down or under attack, this PR implements a flag indicating its status in order to stop Trezor clients from registering their inputs into CoinJoin rounds.

To alleviate any concerns about the affiliate server being brought down, we could have two or more affiliate servers with different ISPs running simultaneously. The coordinator could use any of them to obtain the signed CoinJoin request.

onvej-sl commented on WalletWasabi/Affiliation/Models/AffiliateInformation.cs:
```diff
@@ -0,0 +1,15 @@
+
+public record AffiliateInformation
+(
+	IEnumerable<AffiliationFlag> RunningAffiliateServers,
```
Done in https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/c9a7c51be9ac3347ab9f0da6906512e7144f7f3a. See also https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/65932f3644c2326da985fd94dd24886622edd315 and https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/fe3eccbae0fa182ec2aa8dd023485ef010e4a65e.

onvej-sl commented on WalletWasabi/Affiliation/AffiliateInformationUpdater.cs:
```diff
@@ -0,0 +1,297 @@
+		RemoveHandlers(Arena);
+	}
+
+	public IReadOnlyDictionary<uint256, IReadOnlyDictionary<AffiliationFlag, byte[]>> GetCoinjoinRequests()
```
Done in https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/b8ebc4ec9b3a243a3f7ad8da5085d4b96824b26e.

onvej-sl commented on WalletWasabi/Affiliation/AffiliationManager.cs:
```diff
@@ -0,0 +1,51 @@
+	public AffiliationManager(Arena arena, ImmutableDictionary<AffiliationFlag, string> urls, string privateKeyHex)
+	{
+		Arena = arena;
+		HttpClient httpclient = new();
```
Partial fix is in https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/6cbfa6ec17b955f718aa2ba8efe18b51bd16c4f3.

I briefly read the following articles:
  * https://learn.microsoft.com/en-us/aspnet/core/fundamentals/http-requests?view=aspnetcore-7.0#consumption-patterns
  * https://www.aspnetmonsters.com/2016/08/2016-08-27-httpclientwrong/
  * https://learn.microsoft.com/en-us/dotnet/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests

I think I understand the concept of `IHttpClientFactory` but I don't fully understand how [dependency injection](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/dependency-injection?view=aspnetcore-7.0) works and I need your help here. I don't know how to easily retrieve an instance of `IHttpClientFactory`.

onvej-sl replied:
Do you have any opinion or advice on this?

onvej-sl replied:
Regarding `HttpClient`, see https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/0121ce1c1882d9c7a280e96686fb0912ff1deae1. Nevertheless, I'm still not happy with the solution. [Documentation](https://learn.microsoft.com/en-us/dotnet/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests) says:
> Another issue that developers run into is when using a shared instance of HttpClient in long-running processes. In a situation where the HttpClient is instantiated as a singleton or a static object, it fails to handle the DNS changes as described in this [issue](https://github.com/dotnet/runtime/issues/18348) of the dotnet/runtime GitHub repository.
>
> ...
>
> To address the issues mentioned above and to make HttpClient instances manageable, .NET Core 2.1 introduced two approaches, one of them being [IHttpClientFactory](https://learn.microsoft.com/en-us/dotnet/api/system.net.http.ihttpclientfactory). It's an interface that's used to configure and create HttpClient instances in an app through Dependency Injection (DI). It also provides extensions for Polly-based middleware to take advantage of delegating handlers in HttpClient.
>
> The alternative is to use SocketsHttpHandler with configured PooledConnectionLifetime. This approach is applied to long-lived, static or singleton HttpClient instances. To learn more about different strategies, see [HttpClient guidelines for .NET](https://learn.microsoft.com/en-us/dotnet/fundamentals/networking/http/httpclient-guidelines).

So it still seems to me that we are mixing two approaches together.

onvej-sl replied:
> So what about https://github.com/kiminuo/WalletWasabi/commit/54c70d2fe1ca9ad96b6b7247c8c1fd761d2ee488 ?

This is almost the same as it was before Lucas opened this thread (see https://github.com/zkSNACKs/WalletWasabi/pull/9769#discussion_r1046316807). With the exception that i didn't use `IHttpClientFactory` to create `HttpClient`. But that doesn't make much difference since `AffiliateManager` is initialized only once.

onvej-sl replied:
I still think that we are mixing two approaches but if you are ok with it I'm ok with it as well.

onvej-sl commented on WalletWasabi/Affiliation/RoundData.cs:
```diff
@@ -0,0 +1,113 @@
+	{
+		Inputs = new();
+		RoundParameters = null;
+		Transaction = null;
```
I'm not sure if I understand. Is this what you mean?

I can split `RoundData` in two classes let's say `RoundData` and `RoundDataReady`. `RoundData` is mutable whereas `RoundDataReady` is immutable. `RoundData` implements
```
	public void AddInput(Coin coin, AffiliationFlag affiliationFlag, bool zeroCoordinationFee);
    	public RoundDataReady AddTransaction(NBitcoin.Transaction transaction);
```
whereas `RoundDataReady` implements
```
	public Body GetAffiliationData(AffiliationFlag affiliationFlag)
```
.

onvej-sl replied:
Done in https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/2e2dfc92c411a5f4b97a35c9a5b72e57b9250c98.

onvej-sl commented on WalletWasabi/Affiliation/AffiliateServerStatusUpdater.cs:
```diff
@@ -0,0 +1,96 @@
+			StatusResponse result = await client.GetStatus(new StatusRequest(), cancellationToken);
+			return true;
+		}
+		catch (Exception exception)
```
> Catch only the exception's you are interested on.

I'm interested in this exception. The purpose of this function is nothing else than to determine whether an affiliate is responding.

lontivero replied:
In that case catch the appropriated exception.

onvej-sl replied:
Is this what you mean? https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/c0956ea60df9215b44e33d91051c15aa51a12199

To be honest, it should be in your interest not to publish the affiliation flag if anything goes wrong.

onvej-sl replied:
I reverted https://github.com/zkSNACKs/WalletWasabi/commit/c0956ea60df9215b44e33d91051c15aa51a12199. I still think that it is not completely wrong to catch _all_ exceptions because it is safer not to publish the affiliation flag if _anything_ fails.

> AffiliateServerStatusUpdater.IsAffiliateServerRunningAsync implementation would be much simpler with a for-loop + using an AffiliateServerHttpApiClient.SendAsync overload that is dumb and does not support requestNumber.

There is a functionality that repeats http request if the previous request didn't succeed. If I understand it right, you want to move this functionality from from `AffiliateServerHttpApiClient` to `AffiliateServerStatusUpdater`. I don't think that is right, because this functionality has nothing to do with the `AffiliateServerStatusUpdater`. `AffiliateServerStatusUpdater` should be agnostic to the implementation of `AffiliateServerHttpApiClient`.

nopara73 commented:
Thank you for sharing your concerns. I now understand why you believe the affiliate server needs to confirm the entire transaction. However, the proposed solution goes against the principles of coinjoin, where clients should be able to control their own security without relying on central third parties that they do not get stolen from.

In light of this, to avoid a stalemate, I suggest that we take note of the theoretical attack vector you described and plan to address it in the future when the Trezor device is powerful enough to handle the edge cases you are raising.  
Unfortunately we cannot give out a license for the affiliate server to disrupt our rounds. It is an unacceptable compromise, as it would undermine the reliability of the entire system. Thank you for considering our perspective.

onvej-sl commented on WalletWasabi/Affiliation/AffiliateInformationUpdater.cs:
```diff
@@ -0,0 +1,252 @@
+
+namespace WalletWasabi.Affiliation;
+
+public class CoinjoinRequestsUpdater : PeriodicRunner
```
Fixed in https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/9f29da614fd026e86dac868d26b4c089ff5d69c3.

onvej-sl commented on WalletWasabi/Affiliation/AffiliateInformationUpdater.cs:
```diff
@@ -0,0 +1,252 @@
+public class CoinjoinRequestsUpdater : PeriodicRunner
+{
+	private static TimeSpan AffiliateServerTimeout = TimeSpan.FromSeconds(60);
+	private static TimeSpan Interval = TimeSpan.FromSeconds(1);
```
Done in https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/bb1601451616d277b479208c1c8da224930edce3.

nopara73 reviewed (commented):
Concept ACK assuming there's a kill switch that the Wasabi Coordinator can pull, which signals Trezor clients to not participate in coinjoins in case something is wrong with the Trezor server.

onvej-sl commented on WalletWasabi/Affiliation/AffiliateServerStatusUpdater.cs:
```diff
@@ -0,0 +1,84 @@
+	{
+		try
+		{
+			StatusResponse result = await client.GetStatus(new StatusRequest(), cancellationToken).ConfigureAwait(false);
```
Fixed in https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/44c1d5e4c8b7fce1282c241b6d69921f2d600f70.

onvej-sl replied:
Done in https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/edd71736c829f535098f8b5e618062facb69e77f.

onvej-sl commented on WalletWasabi/Affiliation/AffiliateInformationUpdater.cs:
```diff
@@ -0,0 +1,252 @@
+
+	private void RemoveRounds()
+	{
+		while (RoundsToRemove.TryDequeue(out uint256 roundId))
```
Done in https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/edd71736c829f535098f8b5e618062facb69e77f.

onvej-sl commented on WalletWasabi/Affiliation/AffiliateInformationUpdater.cs:
```diff
@@ -0,0 +1,252 @@
+
+	private async Task UpdateCoinjoinRequestsAsync(CancellationToken cancellationToken)
+	{
+		while (RoundsToUpdate.TryDequeue(out uint256 roundId))
```
Done in https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/edd71736c829f535098f8b5e618062facb69e77f.

onvej-sl commented on WalletWasabi/Affiliation/AffiliateInformationUpdater.cs:
```diff
@@ -0,0 +1,252 @@
+
+	private async Task UpdateCoinjoinRequestsAsync(uint256 roundId, CancellationToken cancellationToken)
+	{
+		if (!RoundData.TryGetValue(roundId, out RoundData roundData))
```
Done in https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/edd71736c829f535098f8b5e618062facb69e77f.

onvej-sl commented:
Thank you all for your help! I'm going on holiday tomorrow for 14 days. That means I will continue with this pull request in the new year!

onvej-sl commented on WalletWasabi/Affiliation/AffiliationFlag.cs:
```diff
@@ -0,0 +1,56 @@
+
+	private static bool IsAlphanumeric(string text)
+	{
+		return text.All(char.IsLetterOrDigit);
```
I wanted to avoid unicode characters. Fixed in https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/077d2e3ef820a10ddc1cc4f44fd5e36f89a575f9.

onvej-sl commented on WalletWasabi/Affiliation/Client.cs:
```diff
@@ -0,0 +1,136 @@
+		  where TRequest : class
+		  where TResponse : class
+	{
+		string requestSerialized = Serialize(request);
```
Fixed in https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/bf977ffaa5bf87c21b98256eb65081e9b8f0a0c4.

onvej-sl commented on WalletWasabi/Affiliation/Client.cs:
```diff
@@ -0,0 +1,136 @@
+		using CancellationTokenSource requestTimeoutCTS = new(requestTimeout);
+		using CancellationTokenSource linkedCTS = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, requestTimeoutCTS.Token);
+
+		return await Client.SendAsync(HttpMethod.Post, GetUriEndPoint(action), content, linkedCTS.Token);
```
Fixed in https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/bf977ffaa5bf87c21b98256eb65081e9b8f0a0c4.

onvej-sl commented on WalletWasabi/Affiliation/Client.cs:
```diff
@@ -0,0 +1,136 @@
+		{
+			try
+			{
+				return await SendAsync(action, jsonString, requestTimeout, cancellationToken);
```
Fixed in https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/bf977ffaa5bf87c21b98256eb65081e9b8f0a0c4.

onvej-sl commented on WalletWasabi/Affiliation/Client.cs:
```diff
@@ -0,0 +1,136 @@
+
+public class AffiliateServerHttpApiClient
+{
+	private IHttpClient Client;
```
Fixed in https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/bf977ffaa5bf87c21b98256eb65081e9b8f0a0c4.

onvej-sl commented on WalletWasabi/Affiliation/CoinjoinRequestsUpdater.cs:
```diff
@@ -0,0 +1,216 @@
+	private ConcurrentDictionary<uint256, RoundData> RoundData { get; }
+	private Queue<FinalizedRoundDataWithRoundId> RoundsToUpdate { get; }
+
+	public override void Dispose()
```
I have no problem adapting to the convention. But I have a few questions:
  * Do you put a `Dispose` method even below private methods?
  * Is CodeFactor OK with that?
  * Are these styling rules documented anywhere?

onvej-sl replied:
I have no problem adapting to the convention. But I have a few questions:
  * Do you put a `Dispose` method even below private methods?
  * Is CodeFactor OK with that?
  * Are these styling rules documented anywhere?

onvej-sl replied:
Fixed in https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/b43d6b4a67fb3cdef759e84737b172ae4a4ebd39.

onvej-sl commented on WalletWasabi/Affiliation/CoinjoinRequestsUpdater.cs:
```diff
@@ -0,0 +1,216 @@
+	private ImmutableDictionary<AffiliationFlag, AffiliateServerHttpApiClient> Clients { get; }
+	private Dictionary<uint256, Dictionary<AffiliationFlag, byte[]>> CoinjoinRequests { get; }
+	private ConcurrentDictionary<uint256, RoundData> RoundData { get; }
+	private Queue<FinalizedRoundDataWithRoundId> RoundsToUpdate { get; }
```
I may be wrong but I don't think so.

`RoundsToUpdate` is used on two places with the following call trees:
  * `CoinjoinRequestsUpdater:UpdateCoinjoinRequestsAsync` <-- `CoinjoinRequestsUpdater:ActionAsync` <-- `PeriodicRunner:ExecuteAsync` <-- `AffiliationManager:ExecuteAsync` <-- `WabiSabiCoordinator:ExecuteAsync`
  *  `CoinjoinRequestsUpdater:AddCoinjoinTransaction` <-- `CoinjoinRequestsUpdater:Arena_CoinjoinTransactionAdded` <-- `Arena:FinalizeTransaction` <-- `Arena:StepOutputRegistrationPhaseAsync` <-- `Arena:ActionAsync` <-- `PeriodicRunner:ExecuteAsync` <-- `WabiSabiCoordinator:ExecuteAsync`

So both the places are called from `WabiSabiCoordinator:ExecuteAsync` and I cannot see any place where a new thread is created.

onvej-sl replied:
> So it seems to me that you have two PeriodicRunners (CoinjoinRequestsUpdater and Arena) that can touch the same data without synchronization as Queue is not synchronized.

This is true but not a problem since both the PeriodicRunners use the same thread. To be specific, they use the thread that is used [here](https://github.com/trezor/WalletWasabi/blob/onvej-sl/affiliate-server/WalletWasabi/WabiSabi/WabiSabiCoordinator.cs#L102-L110).

lontivero replied:
Kiminuo is right here. Long running tasks, even when not created with the appropiated TaskOption, end running in a different and own thread (you can print the thead id and confirm it). In this cases there are two approaches:

* lock shared resources to gain exclusive access (used a lot in wasabi) or,
* use a worker with an inbox channel as synchronization mechanism for receiving messages from the outside.

onvej-sl replied:
The test is pretty convincing. I have to rethink how `CoinjoinRequestsUpdater` works.

onvej-sl replied:
> Queue is [not](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.queue-1?view=net-7.0#thread-safety) thread-safe but it looks to me that you access it from two threads really.

Fixed in https://github.com/zkSNACKs/WalletWasabi/pull/9769.

onvej-sl commented on WalletWasabi/Affiliation/CoinjoinRequestsUpdater.cs:
```diff
@@ -0,0 +1,216 @@
+		ChangePhase(roundId, phase);
+	}
+
+	private void AddHandlers(Arena arena)
```
Fixed in https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/9db26a195d244370376202ed0d84a5a7a5e0c1e0.

onvej-sl commented on WalletWasabi/Affiliation/CoinjoinRequestsUpdater.cs:
```diff
@@ -0,0 +1,216 @@
+		Arena.RoundPhaseChanged += Arena_RoundPhaseChanged;
+	}
+
+	private void RemoveHandlers(Arena arena)
```
Fixed in https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/9db26a195d244370376202ed0d84a5a7a5e0c1e0.

onvej-sl commented on WalletWasabi/Affiliation/Models/CoinjoinRequest/Payload.cs:
```diff
@@ -0,0 +1,25 @@
+	[JsonProperty(PropertyName = "body")]
+	public Body Body { get; }
+
+	public byte[] GetCanonicalSerialization()
+	{
+		return Encoding.ASCII.GetBytes(JsonConvert.SerializeObject(this, CanonicalJsonSerializationOptions.Settings));
+	}
```
Do you mean as a replacement for `OrderedContractResolver`? I can't see any advantage in using `JsonPropertyOrder` instead of `OrderedContractResolver`. Is there any?

onvej-sl replied:
Ok, I see your point.

> PS: I'm fine with OrderedContractResolver but I would add more tests for this (I wrote a comment somewhere with more tests).

I think [this test](https://github.com/trezor/WalletWasabi/blob/onvej-sl/affiliate-server/WalletWasabi.Tests/UnitTests/Affiliation/CanonicalSerializationTests.cs#L29-L36) is sufficient since it tests that the only structure that I need to be canonically serialized is serialized correctly which is equivalent to specifying `[JsonProperty(Order = number)]` for each property.

onvej-sl replied:
> Are you against having this test https://github.com/zkSNACKs/WalletWasabi/pull/9769#discussion_r1059682429 too?

In my opinion, the test doesn't bring anything extra that is relevant compared to the test I mentioned.

> I mean it shows how it behaves wrt JsonProperty attribute which is not obvious imho.

This is already tested in the test I mentioned.

onvej-sl replied:
> But do you honesly expect anybody to read this unformatted blob of characters

No i do not. The blob was generated by the affiliate server and the test checks that both the coordinator and the affiliate-server serialize the request in the same way. This is the only thing I consider to be relevant.

If I were to accept that your test does bring something extra that is relevant compared to my test, I would have to ask why it doesn't test for example sorting of
  * properties encoded in a unicode encoding,
  * nested objects,
  * properties of data types other than `string`, `int`, `decimal`, `bool` and `byte[]`,
  * properties with special characters other than an underscore and at sign (for example a space).

> The goal of tests is to persuade readers that some functionality is correct. 

This is a nice definition but it says nothing unless you specify what the functionality is.

onvej-sl replied:
Are you OK with https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/28668f619142e026eaee035d252ac8e363eb19d0?

onvej-sl replied:
Does it mean this issue is resolved?

onvej-sl commented on WalletWasabi/Affiliation/Serialization/ByteArrayJsonConverter.cs:
```diff
@@ -0,0 +1,23 @@
+	{
+		if (reader.Value is string serialized)
+		{
+			return Encoders.Hex.DecodeData(serialized);
```
Done in https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/898ea4b68cb0c40a0bf885b3a22323f163866adf.

lontivero commented on WalletWasabi/Affiliation/IAffiliationManager.cs:
```diff
@@ -0,0 +1,8 @@
+
+public interface IAffiliationManager
+{
+	public AffiliateInformation GetAffiliateInformation();
```
Yes,

onvej-sl replied:
Done in https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/f3ee611515d30d4fc98dceadecbff5a548bcc700.

lontivero commented on WalletWasabi/Affiliation/Serialization/JsonSerializationOptions.cs:
```diff
@@ -0,0 +1,11 @@
+
+namespace WalletWasabi.Affiliation.Serialization;
+
+public static class JsonSerializationOptions
```
Yes, I agree because even when namespaces are for solving names conflicts, it is always better to avoid the conflicts than solving the conflicts.

onvej-sl replied:
Done in https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/2d6397ca3cbfea0aa31cfd196f3bc90b18f28916 and https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/f909c589de66380b60ecf41b2dcc50bbce37b590.

lontivero commented on WalletWasabi/Affiliation/Serialization/ByteArrayJsonConverter.cs:
```diff
@@ -0,0 +1,23 @@
+
+namespace WalletWasabi.Affiliation.Serialization;
+
+public class ByteArrayJsonConverter : JsonConverter<byte[]>
```
I agree here too.

onvej-sl replied:
Done in https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/1df32b88fe3895e61cd63188006540d092cb6295 and https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/2d6397ca3cbfea0aa31cfd196f3bc90b18f28916.

lontivero commented on WalletWasabi/Affiliation/CoinjoinRequestsUpdater.cs:
```diff
@@ -0,0 +1,216 @@
+	private Arena Arena { get; }
+	private CoinJoinRequestRequestsSigner Signer { get; }
+	private ImmutableDictionary<AffiliationFlag, AffiliateServerHttpApiClient> Clients { get; }
+	private Dictionary<uint256, Dictionary<AffiliationFlag, byte[]>> CoinjoinRequests { get; }
+	private ConcurrentDictionary<uint256, RoundData> RoundData { get; }
+	private Queue<FinalizedRoundDataWithRoundId> RoundsToUpdate { get; }
```
After a quite glance i was not able to find any obvious problem with this as it is. However, it is really fragile and it would be better to have lock to make sure we maintain consistence across the multiple data structures.

onvej-sl replied:
I will consider it. Nevertheless, I'm not really a fan of global locks, since it can lead to unexpected delays or even deadlocks. Which is undesirable as long as `Arena_AffiliationAdded`, `Arena_RoundCreated`, `Arena_CoinjoinTransactionAdded` and `Arena_RoundPhaseChanged` run synchronously (by the way, this can be solved using queues).

I can image that I would have two locks:
  * one for `CoinjoinRequests` (which is a nested `ConcurrentDictionary`) and
  * one for `RoundData` (which is a `ConcurrentDictionary` compounded with `ConcurrentBag`)

lontivero commented on WalletWasabi/Affiliation/CoinjoinRequestsUpdater.cs:
```diff
@@ -0,0 +1,216 @@
+public class CoinjoinRequestsUpdater : PeriodicRunner
+{
+	private static readonly TimeSpan AffiliateServerTimeout = TimeSpan.FromSeconds(60);
+	private static readonly TimeSpan Interval = TimeSpan.FromSeconds(1);
```
I don't see any problem with this as it is, or am I misunderstanding it @kiminuo? I mean, there are no 86400 HTTP requests a day, are they?

onvej-sl replied:
> I would expect an async queue[1](https://github.com/zkSNACKs/WalletWasabi/pull/9769#user-content-fn-1-165384a995b6596bb2900d516246fa50) that waits until it has an element in it. Or maybe that channel that Lucas mentioned can be used in the similar way.

Done in https://github.com/zkSNACKs/WalletWasabi/pull/9769.

An alternative is to use
  * [AsyncQueue](https://learn.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.threading.asyncqueue-1) from `Microsoft.VisualStudio.Threading` or
  * [AsyncProducerConsumerQueue](https://github.com/StephenCleary/AsyncEx/blob/master/src/Nito.AsyncEx.Coordination/AsyncProducerConsumerQueue.cs) from `Nito.AsyncEx`
 
as you mentioned. Since both alternatives require a new dependency and implementation of asynchronous queue using [channels](https://learn.microsoft.com/en-us/dotnet/core/extensions/channels) seems pretty straightforward to me, I dediced to implement it myself.

lontivero commented:
Please not forget about this: https://github.com/zkSNACKs/WalletWasabi/pull/9769#pullrequestreview-1219505582
It is a condition by nopara73 so, we can negotiate all the issues/changes/suggestions except that one.

onvej-sl commented on WalletWasabi/Affiliation/Serialization/CanonicalJsonSerializationOptions.cs:
```diff
@@ -0,0 +1,19 @@
+using System.Linq;
```
Done in https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/bc22b1c825eb3af41e5ffa199bcd96078e5b7613.

onvej-sl commented:
> Please not forget about this: https://github.com/zkSNACKs/WalletWasabi/pull/9769#pullrequestreview-1219505582
It is a condition by nopara73 so, we can negotiate all the issues/changes/suggestions except that one.

This is already implemented. If you decide not to allow trezor users to participate, you can simply remove trezor affiliate server from your configuration and (I assume) restart the coordinator. This will cause the trezor affiliation flag not to appear in the status of your coordinator and no trezor users will participate in the following rounds. We can, of course, agree to trigger it differently, for example without restarting the coordinator. I just wanted to say that this is working.

lontivero commented:
> This is already implemented (sic) you can simply remove trezor affiliate server from your configuration

Okay. I think that's enough, we don't expect to use that anyway but it was a condition.

onvej-sl commented on WalletWasabi/Affiliation/AffiliationManager.cs:
```diff
@@ -0,0 +1,54 @@
+
+public class AffiliationManager : BackgroundService, IAffiliationManager, IDisposable
+{
+	public AffiliationManager(Arena arena, ImmutableDictionary<AffiliationFlag, string> urls, string privateKeyHex, IHttpClientFactory httpClientFactory)
```
Done in https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/8df36229a7e9935f7069f3a84ba302d8cf066918.

onvej-sl commented on WalletWasabi/WabiSabi/Models/Serialization/JsonSerializationOptions.cs:
```diff
@@ -29,6 +29,7 @@ public class JsonSerializationOptions
 				new TimeSpanJsonConverter(),
 				new CoinJsonConverter(),
 				new CoinJoinEventJsonConverter(),
+				new Affiliation.Serialization.AffiliationFlagJsonConverter(),
```
Done in https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/5ee6d10f7439c0c1fd8051716efca0694645890b.

onvej-sl commented on WalletWasabi.Tests/UnitTests/Affiliation/SignatureTests.cs:
```diff
@@ -0,0 +1,26 @@
+
+namespace WalletWasabi.Tests.UnitTests.Affiliation;
+
+public class SignatureTests
```
Done in https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/25dba14fb7c779b990e5cd63cb97330151adf53e.

lontivero commented:
@onvej-sl please make sure to rebase this branch on top of master before Wednesday.

onvej-sl commented on WalletWasabi.Tests/UnitTests/Affiliation/SignatureTests.cs:
```diff
@@ -0,0 +1,26 @@
+		var privateKeyHex = Encoders.Hex.EncodeData(privateKeyBytes);
+		var publicKeyHex = Encoders.Hex.EncodeData(publicKeyBytes);
+		return (privateKeyHex, publicKeyHex);
+	}
```
It can be removed if you want. I just wanted to document somewhere how you can generate `AffiliationMessageSignerKey`.

onvej-sl replied:
Done in https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/e872ed09589f9fd4e5dc6476e41a3d5370aa713f.

onvej-sl commented on WalletWasabi/WabiSabi/WabiSabiCoordinator.cs:
```diff
@@ -45,6 +48,7 @@ public WabiSabiCoordinator(CoordinatorParameters parameters, IRPCClient rpc, ICo
 			transactionArchiver,
 			coinJoinScriptStore,
 			coinVerifier);
+		AffiliationManager = new(Arena, Config.AffiliateServers, Config.AffiliationMessageSignerKey, httpClientFactory);
```
Is there any problem with `AffiliationManager` being instantiated without any URIs?

onvej-sl commented on WalletWasabi/WabiSabi/Backend/WabiSabiConfig.cs:
```diff
@@ -164,6 +166,14 @@ public WabiSabiConfig(string filePath) : base(filePath)
+
+	[DefaultAffiliateServersAttribute]
+	[JsonProperty(PropertyName = "AffiliateServers", DefaultValueHandling = DefaultValueHandling.Populate)]
+	public ImmutableDictionary<AffiliationFlag, string> AffiliateServers { get; set; } = ImmutableDictionary<AffiliationFlag, string>.Empty;
```
What you mean by "this way"?

onvej-sl replied:
I see. This is intentional. First, the URIs are not known yet. Second, they should be kept private.

Kukks commented on WalletWasabi/Affiliation/AsyncQueue.cs:
```diff
@@ -0,0 +1,37 @@
+using System.Threading.Tasks;
+using System.Threading.Channels;
+
+public class AsyncQueue<T>
```
Channels are a great use for this IMO.

Define a channel with the type:

`private readonly Channel<string> PendingIncomingMessages = Channel.CreateUnbounded<string>();`


Write to the channel when you wish to queue an item
`await PendingIncomingMessages.Writer.WriteAsync("hi");`

Have a service that is running that continually reads from the channel. 
```

_ = ProcessChannel(PendingIncomingMessages, HandleIncomingMessage, messageCts.Token);


private Task<bool> HandleIncomingMessage(string message, CancellationToken token)
{
    try {
        Console.Out.WriteLineAsync(message, token);
        return true;
    }
    catch{
    // retry later
    	return false;
    }
}

private async Task ProcessChannel<T>(Channel<T> channel, Func<T, CancellationToken, Task<bool>> processor,
            CancellationToken cancellationToken)
        {
            while (await channel.Reader.WaitToReadAsync(cancellationToken))
            {
                if (channel.Reader.TryPeek(out var evt))
                {
                    var linked = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
                    linked.CancelAfter(5000);
                    if (await processor(evt, linked.Token))
                    {
                        channel.Reader.TryRead(out _);
                    }
                }
            }
        }
}

```

lontivero replied:
Yes, I wouldn't wrap it neither but right now we must focus on correctness. If the code doesn't have a clear problem we can point out then the comment is a suggestion.

onvej-sl replied:
> My experience is that if one takes such a task and it's not done superbly well, then people won't use it and that beats the purpose imo.

Reusability is not the only purpose of class extraction. The purpose may also be for example better readibility.

> If the code doesn't have a clear problem we can point out then the comment is a suggestion.

Do you see any problem in terms of correctness?

molnard commented:
The name `AffiliateServer` is hard to understand for ppl. This is true for the devs as well. Can I propose renaming that to something like `TrezorHardwareWalletAuthorizer`?

andrewkozlik commented:
> The name `AffiliateServer` is hard to understand for ppl. This is true for the devs as well. Can I propose renaming that to something like `TrezorHardwareWalletAuthorizer`?

We tried to design this mechanism so that it would be usable by any affiliate, e.g. BTCPay. In principle it's not limited to Trezor or even to hardware wallets. So if you want to change the name, then I'd recommend using a name that is affiliate-agnostic. Its relationship to Trezor can be better explained in a code comment.

molnard commented:
> In principle it's not limited to Trezor or even to hardware wallets

In this case, it makes less sense to change it.

onvej-sl commented on WalletWasabi/Affiliation/RoundData.cs:
```diff
@@ -0,0 +1,35 @@
+	private ConcurrentBag<AffiliateCoin> Inputs { get; }
+	private RoundParameters RoundParameters { get; }
+
+	public void AddInput(Coin coin, AffiliationFlag affiliationFlag, bool zeroCoordinationFee)
```
This is suggestion is already outdated.

onvej-sl commented on WalletWasabi/Affiliation/AffiliateServerStatusUpdater.cs:
```diff
@@ -0,0 +1,84 @@
+		}
+	}
+
+	private async Task UpdateRunningAffiliateServersAsync(CancellationToken cancellationToken)
+	{
+		await Task.WhenAll(Clients.Select(x => UpdateRunningAffiliateServersAsync(x.Key, x.Value, cancellationToken))).ConfigureAwait(false);
+	}
```
In my opinion, the fact that a method is called from only one place in the code is not in itself a sufficient reason to inline it.

onvej-sl commented on WalletWasabi/Affiliation/RoundData.cs:
```diff
@@ -0,0 +1,35 @@
+		RoundParameters = roundParameters;
+	}
+
+	private ConcurrentBag<AffiliateCoin> Inputs { get; }
```
Done in https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/ff1e895f61cc9766c234d869856f563350570671.

onvej-sl commented on WalletWasabi.Backend/Startup.cs:
```diff
@@ -109,6 +110,12 @@ public void ConfigureServices(IServiceCollection services)
 		});
+		services.AddSingleton(serviceProvider =>
+		{
+			var global = serviceProvider.GetRequiredService<Global>();
```
Done in https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/c187bf38111fe51bbd5b67c101d8eaa6502e910c.

onvej-sl replied:
I had to revert it in https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/e03adc3d8e8b5d702c5d2c6068e96aeda888e507 because it didn't work.

onvej-sl commented on WalletWasabi/Affiliation/RoundData.cs:
```diff
@@ -0,0 +1,78 @@
+			Logging.Logger.LogWarning($"Missing {name} for outpoint {Convert.ToHexString(outpoint.ToBytes())}.");
+		}
+
+		foreach (OutPoint outpoint in pairsOutpoints.Except((IEnumerable<OutPoint>)outpoints))
```
Done in https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/ed19a503d760f8d82e8665d38afcb2524556edb0.

onvej-sl commented on WalletWasabi/Affiliation/CoinjoinRequestsUpdater.cs:
```diff
@@ -0,0 +1,242 @@
+			Body body = finalizedRoundData.GetAffiliationData(affiliationFlag);
+			byte[] result = await GetCoinjoinRequestAsync(affiliateServerHttpApiClient, body, cancellationToken).ConfigureAwait(false);
+
+			if (CoinjoinRequests.TryGetValue(roundId, out ConcurrentDictionary<AffiliationFlag, byte[]> coinjoinRequests))
```
Done in https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/5ba8b483b6fc2056d0e5c361b6f960957e755ef7.

onvej-sl commented on WalletWasabi/Affiliation/CoinjoinRequestsUpdater.cs:
```diff
@@ -0,0 +1,242 @@
+			{
+				try
+				{
+					UpdateCoinjoinRequestsAsync(finalizedRoundDataWithRoundId.RoundId, finalizedRoundDataWithRoundId.FinalizedRoundData, cancellationToken);
```
Fixed in https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/5ba8b483b6fc2056d0e5c361b6f960957e755ef7.

andrewkozlik commented:
Hi everyone, it seems to us that @onvej-sl has addressed all the crucial review comments in this PR. He also rebased this branch on top of master and squashed all the fixup commits, see https://github.com/zkSNACKs/WalletWasabi/pull/10055 where the CI tests should be passing. As we agreed with @MaxHillebrand and discussed with @lontivero and @kiminuo, can we consider the work on this PR complete from the side of SatoshiLabs and hand this over to you? You can add any finishing touches or hit the merge button on #10055 if it's to your satisfaction.

lontivero commented on WalletWasabi/WabiSabi/Backend/Events/InputAddedEventArgs.cs:
```diff
@@ -0,0 +1,19 @@
+
+public class InputAddedEventArgs : EventArgs
+{
+	public InputAddedEventArgs(uint256 roundId, Coin coin, bool isCoordinationFeeExempted) : base()
```
nop, i think it is important to be clear about what fee we are talking about, that's why in all the code we have miningFee and coordinationFee instead of simply fee.

lontivero commented on WalletWasabi/WabiSabi/Models/ReadyToSignRequestRequest.cs:
```diff
@@ -1,5 +1,12 @@
+public record ReadyToSignRequestRequest(
+	uint256 RoundId,
+	Guid AliceId,
+	[property: JsonProperty(DefaultValueHandling = DefaultValueHandling.Populate), DefaultAffiliationFlag()] AffiliationFlag AffiliationFlag
```
It is necessary here too because that value will be sent to the coordinator and this data contract needs to contain it.

lontivero commented on WalletWasabi/Affiliation/AffiliateServerStatusUpdater.cs:
```diff
@@ -0,0 +1,84 @@
+	{
+		try
+		{
+			StatusResponse result = await client.GetStatusAsync(new StatusRequest(), cancellationToken).ConfigureAwait(false);
```
@kiminuo note that this is simply an Unit type that is pretty convenient. I introduced it years ago but other devs didn't like it because they were not familiar with it (UI team uses Unit everywhere). So, we could reintroduce the Unit and replace this and other "empty" classes.

lontivero commented:
We will continue here: https://github.com/zkSNACKs/WalletWasabi/pull/10091
